File: python/app/service/nodetype_service.py
"""
NodeType service implementation.
"""

from typing import List, Tuple

from app.repository import NodeType, NodeTypeRepository, ListOptions, ListResult


class NodeTypeService:
    """NodeType business logic service."""

    def __init__(self, repo: NodeTypeRepository):
        self.repo = repo

    async def create(self, tenant_id: str, name: str, description: str, schema: str) -> NodeType:
        """Create a new node type."""
        if not tenant_id:
            raise ValueError("tenant_id is required")
        if not name:
            raise ValueError("name is required")

        node_type = NodeType(
            tenant_id=tenant_id,
            name=name,
            description=description,
            schema=schema,
        )
        return await self.repo.create(node_type)

    async def get_by_id(self, tenant_id: str, id: str) -> NodeType:
        """Retrieve a node type by ID."""
        if not id:
            raise ValueError("id is required")
        if not tenant_id:
            raise ValueError("tenant_id is required")
        return await self.repo.get_by_id(tenant_id, id)

    async def update(self, tenant_id: str, id: str, name: str, description: str, schema: str) -> NodeType:
        """Update an existing node type."""
        if not id:
            raise ValueError("id is required")
        if not tenant_id:
            raise ValueError("tenant_id is required")

        node_type = await self.repo.get_by_id(tenant_id, id)

        if name:
            node_type.name = name
        if description:
            node_type.description = description
        if schema:
            node_type.schema = schema

        return await self.repo.update(node_type)

    async def delete(self, tenant_id: str, id: str) -> None:
        """Delete a node type."""
        if not id:
            raise ValueError("id is required")
        if not tenant_id:
            raise ValueError("tenant_id is required")
        await self.repo.delete(tenant_id, id)

    async def list(self, tenant_id: str, page_size: int, page_token: str) -> Tuple[List[NodeType], ListResult]:
        """Retrieve node types with pagination."""
        if not tenant_id:
            raise ValueError("tenant_id is required")

        opts = ListOptions(page_size=page_size, page_token=page_token)
        return await self.repo.list(tenant_id, opts)

File: go/internal/service/nodetype_service.go
package service

import (
	"context"
	"fmt"

	"github.com/hemanthpathath/flex-db/go/internal/repository"
)

// NodeTypeService handles node type business logic
type NodeTypeService struct {
	repo repository.NodeTypeRepository
}

// NewNodeTypeService creates a new NodeTypeService
func NewNodeTypeService(repo repository.NodeTypeRepository) *NodeTypeService {
	return &NodeTypeService{repo: repo}
}

// Create creates a new node type
func (s *NodeTypeService) Create(ctx context.Context, tenantID, name, description, schema string) (*repository.NodeType, error) {
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	if name == "" {
		return nil, fmt.Errorf("name is required")
	}

	nodeType := &repository.NodeType{
		TenantID:    tenantID,
		Name:        name,
		Description: description,
		Schema:      schema,
	}

	return s.repo.Create(ctx, nodeType)
}

// GetByID retrieves a node type by ID
func (s *NodeTypeService) GetByID(ctx context.Context, tenantID, id string) (*repository.NodeType, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	return s.repo.GetByID(ctx, tenantID, id)
}

// Update updates an existing node type
func (s *NodeTypeService) Update(ctx context.Context, tenantID, id, name, description, schema string) (*repository.NodeType, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}

	nodeType, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		return nil, err
	}

	if name != "" {
		nodeType.Name = name
	}
	if description != "" {
		nodeType.Description = description
	}
	if schema != "" {
		nodeType.Schema = schema
	}

	return s.repo.Update(ctx, nodeType)
}

// Delete deletes a node type
func (s *NodeTypeService) Delete(ctx context.Context, tenantID, id string) error {
	if id == "" {
		return fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return fmt.Errorf("tenant_id is required")
	}
	return s.repo.Delete(ctx, tenantID, id)
}

// List retrieves node types with pagination
func (s *NodeTypeService) List(ctx context.Context, tenantID string, pageSize int32, pageToken string) ([]*repository.NodeType, *repository.ListResult, error) {
	if tenantID == "" {
		return nil, nil, fmt.Errorf("tenant_id is required")
	}

	opts := repository.ListOptions{
		PageSize:  int(pageSize),
		PageToken: pageToken,
	}
	return s.repo.List(ctx, tenantID, opts)
}
